packages/wdio-cli/src/launcher.js

```diff
@@ -55,49 +55,61 @@ class Launcher {
      */
     async run () {
-        let config = this.configParser.getConfig()
-        let caps = this.configParser.getCapabilities()
-        const launcher = initialiseServices(config, caps, 'launcher')
-
-        /**
-         * run pre test tasks for runner plugins
-         * (e.g. deploy Lambda function to AWS)
-         */
-        await this.runner.initialise()
-
-        /**
-         * run onPrepare hook
-         */
-        log.info('Run onPrepare hook')
-        await runOnPrepareHook(config.onPrepare, config, caps)
-        await runServiceHook(launcher, 'onPrepare', config, caps)
-
         /**
          * catches ctrl+c event
          */
         exitHook(::this.exitHandler)
+        let exitCode
+        let onCompleteResults
 
-        let exitCode = await this.runMode(config, caps)
+        try {
+            const config = this.configParser.getConfig()
+            const caps = this.configParser.getCapabilities()
+            const launcher = initialiseServices(config, caps, 'launcher')
 
-        /**
-         * run onComplete hook
-         * even if it fails we still want to see result and end logger stream
-         */
-        log.info('Run onComplete hook')
-        await runServiceHook(launcher, 'onComplete', exitCode, config, caps)
+            /**
+             * run pre test tasks for runner plugins
+             * (e.g. deploy Lambda function to AWS)
+             */
+            await this.runner.initialise()
 
-        const onCompleteResults = await runOnCompleteHook(config.onComplete, config, caps, exitCode, this.interface.result)
+            /**
+             * run onPrepare hook
+             */
+            log.info('Run onPrepare hook')
+            await runOnPrepareHook(config.onPrepare, config, caps)
+            await runServiceHook(launcher, 'onPrepare', config, caps)
+
+            exitCode = await this.runMode(config, caps)
+
+            /**
+             * run onComplete hook
+             * even if it fails we still want to see result and end logger stream
+             */
+            log.info('Run onComplete hook')
+            await runServiceHook(launcher, 'onComplete', exitCode, config, caps)
 
-        // if any of the onComplete hooks failed, update the exit code
-        exitCode = onCompleteResults.includes(1) ? 1 : exitCode
+            onCompleteResults = await runOnCompleteHook(config.onComplete, config, caps, exitCode, this.interface.result)
 
-        await logger.waitForBuffer()
+            // if any of the onComplete hooks failed, update the exit code
+            exitCode = onCompleteResults.includes(1) ? 1 : exitCode
 
-        this.interface.finalise()
-        return exitCode
+            await logger.waitForBuffer()
+
+            this.interface.finalise()
+            return exitCode
+        } catch (err) {
+            if (!onCompleteResults) {
+                /**
+                 * `shutdown` event is added by `exitHook`
```

Comment:
Do you mean is caught by `exitHook`?

Reply:
it is defined few lines above, see https://github.com/webdriverio/webdriverio/pull/4332/files#diff-03f312cc95c9f2a6588676367c2fcbe8R64